File: JavaAlgorithm/src/main/java/algo/tzashinorpu/SpecialSubject/BinaryTree/sufficientSubset_1080.java
package algo.tzashinorpu.SpecialSubject.BinaryTree;

import algo.tzashinorpu.SpecialSubject.TreeNode;

public class sufficientSubset_1080 {
    /*
    if root.left == root.right == null, root is leaf with no child {
    if root.val < limit, we return null;
    else we return root.
    }
    if root.left != null, root has left child {
        Recursively call sufficientSubset function on left child,
        limit - root.val 当前节点有效的最小值
        with limit = limit - root.val
    }
    if root.right != null, root has right child {
        Recursively call sufficientSubset function on right child,
        with limit = limit - root.val
    }
    if root.left == root.right == null, -> root.left == root.right 只有在两边都是 null 的时候才相等
    root has no more children, no valid path left, --> 如果有一个节点不是 null 说明又一条有效的路径被保留了
    we return null;
    else we return root.
     */
    public TreeNode sufficientSubset(TreeNode root, int limit) {
        if (root == null)
            return null;
        if (root.left == null && root.right == null)
            return root.val < limit ? null : root;
        root.left = sufficientSubset(root.left, limit - root.val);
        root.right = sufficientSubset(root.right, limit - root.val);
        return root.left == root.right ? null : root;
    }
}

File: Clion/ZeroTrac/Tree/sufficientSubset_1080.cpp
//
// Created by TzashiNorpu on 10/24/2022.
//
#include "tree.h"

using namespace Tree;

TreeNode *Solution::sufficientSubset(TreeNode *root, int limit) {
  if (!root)return nullptr;
  // 前序位置，接收父节点传递的 limit 约束决定叶子结点是否需要被删除
  if (!root->left && !root->right) return root->val < limit ? nullptr : root;
  // 先对左右子树进行删除，接收返回值
  root->left = sufficientSubset(root->left, limit - root->val);
  root->right = sufficientSubset(root->right, limit - root->val);
  // 后序位置，根据子树的删除情况决定自己是否需要被删除
  // 如果左右子树不满足 limit - root.val 的约束，那么就存在经过 root
  // 节点的路径不满足约束，也就说明 root 节点是「不足节点」，需要被删掉
  return root->left == root->right ? nullptr : root;
}